File: maximum-subarray/maximum-subarray.js
/**
 * nums(n) = Math.max(nums[i], nums[i] + maxSubArray(i+1))
 * Time O(n) we hit every index once
 * Space O(n) we can reuse the nums array
 */
var maxSubArray = function(nums) {
    let max = -Infinity;
    const dp = (i) => {
        if (i >= nums.length) return 0;
        nums[i] = Math.max(nums[i], nums[i] + dp(i+1))
        max = Math.max(max, nums[i])
        return nums[i]
    }
    dp(0)
    return max
};

File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 0) return 0;
        int result = nums[0];
        int[] dp = new int[nums.length+1];
        dp[0] = nums[0];

        for (int i = 1; i < nums.length; i++) {
            nums[i] = Math.max(nums[i], nums[i] + nums[i-1]);
            result = Math.max(result, nums[i]);
        }
        
        return result;
    }
}